File: solutions/0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
private:
    class rect {
    private:
        int index;
        int height;
    public:
        void setIndex(int ind){
            index = ind;
        } 
        void setHeight(int hei){
            height = hei;
        }
        int getIndex() {return index;}
        int getHeight() {return height;}

    };
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<rect> st;
        int maxArea = 0;
        for(int i = 0; i < heights.size(); i++){
            int lastIndex = i;
            while(!st.empty() && st.top().getHeight() > heights[i]){
                int currArea = st.top().getHeight() * (i - st.top().getIndex());
                maxArea = max(currArea, maxArea);
                lastIndex = st.top().getIndex();
                st.pop();
            }
            rect newRect;
            newRect.setIndex(lastIndex);
            newRect.setHeight(heights[i]);
            st.push(newRect);
        }

        int i = heights.size();
        while(!st.empty()){
            int currArea = st.top().getHeight() * (i - st.top().getIndex());
            maxArea = max(currArea, maxArea);
            st.pop();
        }
        return maxArea;
    }
};

File: solutions/0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        heights.append(0)
        stack = []
        max_area = 0
        for i in range(len(heights)):
            while stack and heights[i] < heights[stack[-1]]:
                h = stack.pop()
                width = i
                if stack:
                    width = width - stack[-1] - 1
                area = heights[h] * width
                max_area = max(max_area, area)
            stack.append(i)
        return max_area